File: src/backtracking/cpp/hamiltonian_paths/test/test_hamiltonian_paths.cpp
#include "hamiltonian_paths.h"
#include "gtest/gtest.h"

TEST(HamiltonianPathsTest, TestPositive1) {
  Graph G;
  G.add_edge(0, 1);
  G.add_edge(0, 2);
  G.add_edge(0, 3);
  G.add_edge(1, 2);
  G.add_edge(1, 3);
  G.add_edge(2, 3);

  std::vector<std::vector<int>> expected = {
      {0, 1, 2, 3}, {0, 1, 3, 2}, {0, 2, 1, 3}, {0, 2, 3, 1}, {0, 3, 1, 2},
      {0, 3, 2, 1}, {1, 0, 2, 3}, {1, 0, 3, 2}, {1, 2, 0, 3}, {1, 2, 3, 0},
      {1, 3, 0, 2}, {1, 3, 2, 0}, {2, 0, 1, 3}, {2, 0, 3, 1}, {2, 1, 0, 3},
      {2, 1, 3, 0}, {2, 3, 0, 1}, {2, 3, 1, 0}, {3, 0, 1, 2}, {3, 0, 2, 1},
      {3, 1, 0, 2}, {3, 1, 2, 0}, {3, 2, 0, 1}, {3, 2, 1, 0},
  };

  std::vector<std::vector<int>> actual = find_hamiltonian_paths(G);
  EXPECT_EQ(expected.size(), actual.size());

  for (auto configuration : actual) {
    EXPECT_TRUE(std::find(expected.begin(), expected.end(), configuration) !=
                expected.end());
  }
}

TEST(HamiltonianPathsTest, TestNegative1) {
  Graph G;
  G.add_edge(0, 2);
  G.add_edge(1, 1);
  G.add_edge(2, 3);
  G.add_edge(2, 4);
  G.add_edge(4, 5);

  std::vector<std::vector<int>> expected = {};
  std::vector<std::vector<int>> actual = find_hamiltonian_paths(G);
  EXPECT_EQ(expected, actual);
}

File: src/backtracking/python/hamiltonian_paths/tests/test_hamiltonian_paths.py
import unittest
from src.hamiltonian_paths import Graph, find_hamiltonian_paths


class TestHamiltonianPaths(unittest.TestCase):
    def test_positive_1(self):
        G = Graph()
        G.add_edge(0, 1)
        G.add_edge(0, 2)
        G.add_edge(0, 3)
        G.add_edge(1, 2)
        G.add_edge(1, 3)
        G.add_edge(2, 3)

        expected = [
            [0, 1, 2, 3],
            [0, 1, 3, 2],
            [0, 2, 1, 3],
            [0, 2, 3, 1],
            [0, 3, 1, 2],
            [0, 3, 2, 1],
            [1, 0, 2, 3],
            [1, 0, 3, 2],
            [1, 2, 0, 3],
            [1, 2, 3, 0],
            [1, 3, 0, 2],
            [1, 3, 2, 0],
            [2, 0, 1, 3],
            [2, 0, 3, 1],
            [2, 1, 0, 3],
            [2, 1, 3, 0],
            [2, 3, 0, 1],
            [2, 3, 1, 0],
            [3, 0, 1, 2],
            [3, 0, 2, 1],
            [3, 1, 0, 2],
            [3, 1, 2, 0],
            [3, 2, 0, 1],
            [3, 2, 1, 0],
        ]

        actual = find_hamiltonian_paths(G)
        self.assertEqual(len(expected), len(actual))

        for configuration in actual:
            self.assertTrue(configuration in expected)

    def test_negative_1(self):
        G = Graph()
        G.add_edge(0, 2)
        G.add_edge(1, 1)
        G.add_edge(2, 3)
        G.add_edge(2, 4)
        G.add_edge(4, 5)

        expected = []
        actual = find_hamiltonian_paths(G)
        self.assertEqual(expected, actual)


if __name__ == "__main__":
    unittest.main()
